parcels/fieldfilebuffer.py

```diff
@@ -14,20 +14,232 @@
 
 
 class _FileBuffer(object):
-    def __init__(self, filename, dimensions, indices, netcdf_engine, timestamp=None,
+    def __init__(self, filename, dimensions, indices, timestamp=None,
                  interp_method='linear', data_full_zdim=None, **kwargs):
         self.filename = filename
         self.dimensions = dimensions  # Dict with dimension keys for file data
         self.indices = indices
         self.dataset = None
-        self.netcdf_engine = netcdf_engine
         self.timestamp = timestamp
         self.ti = None
         self.interp_method = interp_method
         self.data_full_zdim = data_full_zdim
 
 
+class XarrayFileBuffer(_FileBuffer):
+    def __init__(self, *args, **kwargs):
+        super(XarrayFileBuffer, self).__init__(*args, **kwargs)
+
+    def __enter__(self):
+        self.dataset = self.filename
+        return self
+
+    def __exit__(self, type, value, traceback):
+        self.close()
+
+    def close(self):
+        pass
+
+
 class NetcdfFileBuffer(_FileBuffer):
+    def __init__(self, *args, **kwargs):
+        self.lib = np
+        self.netcdf_engine = kwargs.pop('netcdf_engine', 'netcdf4')
+        self.lock_file = kwargs.pop('lock_file', True)
+        super(NetcdfFileBuffer, self).__init__(*args, **kwargs)
+
+    def __enter__(self):
+        try:
+            # Unfortunately we need to do if-else here, cause the lock-parameter is either False or a Lock-object
+            # (which we would rather want to have being auto-managed).
+            # If 'lock' is not specified, the Lock-object is auto-created and managed bz xarray internally.
+            if self.lock_file:
+                self.dataset = xr.open_dataset(str(self.filename), decode_cf=True, engine=self.netcdf_engine)
+            else:
+                self.dataset = xr.open_dataset(str(self.filename), decode_cf=True, engine=self.netcdf_engine, lock=False)
+            self.dataset['decoded'] = True
+        except:
+            logger.warning_once("File %s could not be decoded properly by xarray (version %s).\n         "
+                                "It will be opened with no decoding. Filling values might be wrongly parsed."
+                                % (self.filename, xr.__version__))
+            if self.lock_file:
+                self.dataset = xr.open_dataset(str(self.filename), decode_cf=False, engine=self.netcdf_engine)
+            else:
+                self.dataset = xr.open_dataset(str(self.filename), decode_cf=False, engine=self.netcdf_engine, lock=False)
```

Comment:
the locking only makes sense when one uses Dask, so in the 'regular' NetCDF file buffer, the locking can (or: shall) go. Conversely, when using a DaskFileBuffer, the locking needs to be done - hence the 'except' branch would throw the error then.

Reply:
Fixed in cc794eec3d2a2bf1b479dccf6d8884c45e987f9d